core/session.ts

```diff
@@ -806,3 +816,71 @@ class ActionIndex {
     return this.byName.get(target.name) || [];
   }
 }
+
+
+class ActionAssertionMap {
```

Comment:
An actions index already exists
https://github.com/dataform-co/dataform/blob/754d582f4f588b3c812c40a38ec839607ec3d4f0/core/session.ts#L324 before we start compiling graph chunks https://github.com/dataform-co/dataform/blob/754d582f4f588b3c812c40a38ec839607ec3d4f0/core/session.ts#L344

Why can't that be used to retrieve the action as its fully qualified name, for comparing targets?

Reply:
These might look quit similar because I used code from ActionMap for look up. However, these are two different things. 
ActionMap converts a simple list into a lookup list where each actions is mapped alongs its own target.
On the other hand ActionAssertionMap hold the information of assertions related to each action.
this information is updated in assertions and used later to find the assertions related to any actions. The implementation and usage for these two are very different.

Reply:
As far I can see they only differ in the target type (`Action` vs `ITarget`), the actual code logic is the same. Can you just use `Map<string, dataform.ITarget[]>` and call `targetAsReadableString` to get a fully qualified name for an assertion target key? An alternative way is to create a generic class `ActionMap<ValueType>` and then instantiate it with `ActionMap<Action>` for `indexedActions` and `ActionMap<ITarget>` for `actionAssertionMap`.